framework/http/MessageTrait.php

```diff
@@ -0,0 +1,301 @@
+<?php
+/**
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+
+namespace yii\http;
+
+use Psr\Http\Message\StreamInterface;
+use yii\di\Instance;
+
+/**
+ * MessageTrait provides set of methods to satisfy [[\Psr\Http\Message\MessageInterface]].
+ *
+ * This trait should be applied to descendant of [[\yii\base\BaseObject]] implementing [[\Psr\Http\Message\MessageInterface]].
+ *
+ * @property string $protocolVersion HTTP protocol version as a string.
+ * @property string[][] $headers the message's headers.
+ * @property StreamInterface $body the body of the message.
+ * @property HeaderCollection $headerCollection The header collection. This property is read-only.
+ *
+ * @author Paul Klimov <klimov.paul@gmail.com>
+ * @since 2.1.0
+ */
+trait MessageTrait
+{
+    /**
+     * @var string HTTP protocol version as a string.
+     */
+    private $_protocolVersion;
+    /**
+     * @var HeaderCollection header collection, which is used for headers storage.
+     */
+    private $_headerCollection;
+    /**
+     * @var StreamInterface the body of the message.
+     */
+    private $_body;
+
+
+    /**
+     * Retrieves the HTTP protocol version as a string.
+     * @return string HTTP protocol version.
+     */
+    public function getProtocolVersion()
+    {
+        if ($this->_protocolVersion === null) {
+            $this->_protocolVersion = $this->defaultProtocolVersion();
+        }
+        return $this->_protocolVersion;
+    }
+
+    /**
+     * Specifies HTTP protocol version.
+     * @param string $version HTTP protocol version
+     */
+    public function setProtocolVersion($version)
+    {
+        $this->_protocolVersion = $version;
+    }
+
+    /**
+     * Return an instance with the specified HTTP protocol version.
+     *
+     * @param string $version HTTP protocol version
+     * @return static
+     */
+    public function withProtocolVersion($version)
+    {
+        $this->setProtocolVersion($version);
+        return $this;
```

Comment:
According to PSR, request and response methods are immutable i.e. it should be something like the following:

```php
public function withProtocolVersion($version)
{
    $newThis = clone $this;
    $newThis->setProtocolVersion($version);
    return $newThis;
}
```